pkg/controller/clusterautoscaler/clusterautoscaler.go

```diff
@@ -71,6 +72,11 @@ func AutoscalerArgs(ca *v1.ClusterAutoscaler, cfg *Config) []string {
 		args = append(args, v)
 	}
 
+	if ca.Spec.MaxNodeProvisionTime != nil {
+		v := MaxNodeProvisionTimeArg.Value(*s.MaxNodeProvisionTime)
+		args = append(args, v)
+	}
```

Comment:
Will this break the autoscaler if `*ca.Spec.MaxNodeProvisionTime == ""`? Is there actually any benefit to the field being a string pointer?

Reply:
This is consistent with most other fields in the spec - I think they are defined as pointers deliberately, or the `omitempty` wouldn't work - golang would give a default value for each field vs a nil pointer.

Reply:
https://www.sohamkamani.com/golang/2018-07-19-golang-omitempty/ has a nice explanation on this - AFAICS the user can't pass `""` because that's treated as a nil interface value when marshalling the yaml and specifying `omitempty`

Reply:
I don't think that's the case for string values, if `""` is not a value we are allowing, then because that is the zero value, we can use that as the unset case

See this playground example: https://play.golang.org/p/YaETDGcJsX3, which demonstrates that with a string pointer, you can end up with a pointer to an empty string, which would then get passed through and result in the empty string value, this example also demonstrates that the empty string reference marshalling for omitempty works as well, see on both examples how the field isn't marshalled when printed out

Reply:
In each of the other fields in this spec, there is a clear distinction in meaning between `nil` and the zero value, eg, `*bool` is commonly used when you want to allow a boolean to default to true, by seeing that it is nil, you can default it, otherwise you have no way to no if the zero value (`false`) is there because the user set it, or if it's just because it's the zero value

Reply:
@JoelSpeed Ah I see now, thanks for clarifying.

I think the source of my confusion is that in ScaleDownConfig the values all appear to have the same problem - the interface is similar so I just copied the existing pattern:

```
$ grep "\*string"  pkg/apis/autoscaling/v1/clusterautoscaler_types.go 
	MaxNodeProvisionTime *string `json:"maxNodeProvisionTime,omitempty"`
	DelayAfterAdd *string `json:"delayAfterAdd,omitempty"`
	DelayAfterDelete *string `json:"delayAfterDelete,omitempty"`
	DelayAfterFailure *string `json:"delayAfterFailure,omitempty"`
	UnneededTime *string `json:"unneededTime,omitempty"`

$ grep -e DelayAfter -e UnneededTime  pkg/controller/clusterautoscaler/clusterautoscaler.go | grep nil
	if sd.DelayAfterAdd != nil {
	if sd.DelayAfterDelete != nil {
	if sd.DelayAfterFailure != nil {
	if sd.UnneededTime != nil {
```

If you agree these should be modified in a similar way I can push a followup (and the regex could be changed to be consistent with what's been discussed here?).

Reply:
Yeah I think we should probably look into those other ones as well, as they are string pointers, there's a possibility we could have a bug where a user enters the empty string accidentally and we add a flag such as `--unneededTime= ` which would cause the autoscaler to throw a flag parse error on startup